src/github.com/matrix-org/dendrite/syncapi/consumers/roomserver.go

```diff
@@ -177,6 +202,63 @@ func (s *OutputRoomEvent) lookupStateEvents(
 	return result, nil
 }
 
+func (s *OutputRoomEvent) updateMemberEvent(
+	event gomatrixserverlib.Event, keyID gomatrixserverlib.KeyID,
+	privateKey []byte,
+) (gomatrixserverlib.Event, error) {
+	if event.Type() != "m.room.member" {
+		return event, nil
+	}
+	membership, err := event.Membership()
+	if err != nil {
+		return event, err
+	}
+	if membership != "join" {
+		return event, nil
+	}
+
+	prevEvent, err := s.db.GetMembershipEvent(event.RoomID(), *event.StateKey())
+	if err != nil {
+		return event, err
+	}
+
+	if prevEvent == nil {
+		return event, nil
+	}
+
+	builder := gomatrixserverlib.EventBuilder{
+		Sender:     event.Sender(),
+		RoomID:     event.RoomID(),
+		Type:       event.Type(),
+		StateKey:   event.StateKey(),
+		PrevEvents: event.PrevEvents(),
+		AuthEvents: event.AuthEvents(),
+		Redacts:    event.Redacts(),
+		Depth:      event.Depth(),
+		Content:    event.Content(),
+	}
+
+	var content events.MemberContent
+	if err := json.Unmarshal(prevEvent.Content(), &content); err != nil {
+		return event, err
+	}
+
+	prev := prevMembership{
+		PrevContent: content,
+		PrevID:      prevEvent.EventID(),
+		UserID:      prevEvent.Sender(),
+	}
+
+	if err = builder.SetUnsigned(prev); err != nil {
+		return event, err
+	}
```

Comment:
I'm adding a way to set the unsigned field of an event without building a new one in matrix-org/gomatrixserverlib#55. You will want to use that when it lands.